src/js/storaje.js

```diff
@@ -0,0 +1,68 @@
+const { getState, replaceWholeState } = require("./appState");
+const Recipe = require("./model/Recipe");
+const Ingredient = require("./model/Ingredient");
+const Rating = require("./model/Rating");
+const User = require("./model/User");
+
+function saveStateToStorage() {
+	const STATE_JSON = JSON.stringify(getState);
+	localStorage.setItem("state", STATE_JSON);
+}
+
+function saveInitialStateToStorage() {
+	const INGREDIENTS = [
+		{ ingredient: new Ingredient(1, "macarrones"), quantity: "500g" },
+		{ ingredient: new Ingredient(2, "carne picada"), quantity: "500g" },
+		{ ingredient: new Ingredient(3, "tomate frito"), quantity: "500g" },
+		{ ingredient: new Ingredient(4, "cebolla"), quantity: "1" },
+		{ ingredient: new Ingredient(5, "ajo"), quantity: "4 dientes" }];
+	const STEPS = [
+		"Cocemos los macarrones. Para ello, los introducimos en una cazuela con abundante agua hirviendo. Añadimos un buen puñadito de sal y los cocemos durante unos 12-15 minutos. Hasta que estén al dente. Removemos de vez en cuando, sobre todo al principio, para que no se peguen al fondo de la olla",
+		"Mientras se cocinan, podemos ir preparando la salsa. En una sartén amplia, ponemos a calentar un chorrito de aceite a fuego medio. Después añadimos la cebolla y los ajos picados muy finos. Salpimentamos y removemos con una cuchara de palo mientras cocinamos durante 5 minutos",
+		"Pasado el tiempo, añadimos la carne picada. la salpimentamos y la desmenuzamos con la cuchara de palo para que no queden mazacotes grandes y quede bien suelta. Subimos la fuerza del fuego y la dejamos cocinar hasta que coja algo de color",
+		"Una vez hecho esto, añadimos el tomate frito y espolvoreamos con orégano seco. Mezclamos todos los ingredientes y dejamos cocinar 10 minutos más.",
+		"Cuando los macarrones estén al dente, reservamos uno 150ml del caldo de la cocción en un vaso y los escurrimos del resto del agua",
+		"g Cuando tanto los macarrones como la salsa estén cocinados, añadimos los macarrones a la cazuela donde esté la salsa. Añadimos el caldo de la cocción (que potenciará todos los sabores) y lo mezclamos todo bien, mientras cocinamos todo durante un par de minutos"];
+	const RECIPES = [new Recipe(
+		1,
+		"poderosa",
+		INGREDIENTS,
+		STEPS,
+		"mexicana",
+		"china",
+		1
+	)];
+	const RATINGS = [new Rating(
+		1,
+		"Me encanta",
+		10,
+		1,
+		1
+	)];
+	const USERS = [new User(1, "Pablo")];
+	const STATE = Object({
+		recipes: RECIPES,
+		ratings: RATINGS,
+		users: USERS
+	});
+	localStorage.setItem("state", JSON.stringify(STATE));
+	replaceWholeState(STATE);
+}
+
+function loadStateFromStorage() {
+	if (localStorage.getItem("state") !== null) {
+		const RECIPES = JSON.parse(localStorage.getItem("recipes"));
+		const RATINGS = JSON.parse(localStorage.getItem("ratings"));
+		const USERS = JSON.parse(localStorage.getItem("users"));
+		const STATE = Object({
```

Comment:
Aquí te falta un pequeño detalle (se aplica a los tres arrays). Como los elementos han pasado por `JSON.parse`, han perdido su identidad de clase. Es decir, son objetos normales, no instancias de clase.

Si `Recipe` tuviera un método `calculatePrice`, que calcula el precio de todos los ingredientes (por ejemplo), no podrías hacer `STATE.recipes[0].calculatePrice()`, porque `STATE.recipes[0]` es un `Object` pero no un `Recipe`.

Hay dos formas de abordar este problema. La más intuitiva es crear una nueva receta a partir de la extraida del estado:

```js
const data = STATE.recipes[0]
const recipe = new Recipe(
    data.idRecipe,
    data.name,
    data.ingredients,
    data.steps,
    data.typeOfFood,
    data.origin,
    data.idAuthor)

// Esto lo realizaríamos dentro de un bucle, convirtiendo todas las recetas
// Y así para los usuarios y las valoraciones
```

Un coñazo, ¿no? Hay otra forma más rápida, pero también más difícil de entender, y es más "the javascript way". Es [Object.setPrototypeOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf). Básicamente con esto estás diciendo a Javascript: "este objeto que tengo aquí, aunque tú no te lo creas, es de la clase que te estoy diciendo, así que quiero que tenga todos sus métodos."

Aquí tienes un breve ejemplo:

```js
class Clase { constructor(param) { this.param = param } multiplicar() { return this.param * 2 } }
> undefined
const a = new Clase(5)
> undefined
a.multiplicar()
> 10
const b = { param: 3 }
> undefined
b.multiplicar()
> VM351:1 Uncaught TypeError: b.multiplicar is not a function
Object.setPrototypeOf(b, Clase.prototype)
> Clase {param: 3}
b.multiplicar()
> 6
```

Esto habría que hacerlo en todos los elementos de ratings, users y recipes.